lib/changefeed.js

```diff
@@ -0,0 +1,154 @@
+            log: req.log.child({
+                component: 'changefeed',
+                // Otherwise you'll see a ton of noise from here
+                level: 'fatal'
+            }),
+            url: req.config.vmapi.url,
+            instance: req.config.instanceUuid,
+            service: req.config.serviceName,
+            changeKind: {
+                resource: 'vm',
+                subResources: [
+                    'alias',
+                    'customer_metadata',
+                    'destroyed',
+                    'nics',
+                    'owner_uuid',
+                    'server_uuid',
+                    'state',
+                    'tags'
+                ]
+            }
+        });
+
+        this.vmapiFeed.register();
+        this.vmapiFeed.on('bootstrap', () => {
+            req.log.trace('Changefeed bootstrap complete');
+        });
+
+        const cfDrain = () => {
+            let changeItem;
+            while ((changeItem = this.vmapiFeed.read()) !== null) {
+                // Workaround for duplicated messages received:
+                if (this.latestMsgTimestamp === changeItem.published) {
+                    return;
+                }
+                this.latestMsgTimestamp = changeItem.published;
+                processChange(changeItem);
+            }
+        };
+
+        this.vmapiFeed.on('readable', cfDrain);
+
+        this.vmapiFeed.on('error', (listenerErr) => {
+            req.log.error({err: listenerErr}, 'LISTENER ERROR');
+        });
+
+        this.vmapiFeed.on('connection-end', () => {
+            req.log.debug('Changefeed connection-end');
+        });
+
+        // Handle Incoming Upgrade Requests
+        if (!res.claimUpgrade) {
+            next(new Error('Connection Must Upgrade For WebSockets'));
+            return;
+        }
+
+        const reqdom = process.domain;
+
+        if (reqdom && reqdom.domain) {
+            reqdom.exit();
+        }
+
+        const upgrade = res.claimUpgrade();
+        const ws = shed.accept(req, upgrade.socket, upgrade.head);
+
+        const closeWsObj = () => {
+            if (ws !== undefined) {
+                ws.end();
+            }
+        };
+
+        ws.on('connectionReset', closeWsObj);
+
+        function processChange(changeItem) {
+            const rObj = changeItem.resourceObject;
+            if (rObj && rObj.owner_uuid === req.account.uuid) {
+                const wsMsg = Object.assign({}, changeItem);
+                wsMsg.resourceObject = translate(
+                    wsMsg.resourceObject, req);
+                // translateState in translate in machines.js
+                // changes the down, and stopped states to stopped.
+                // if we don't keep the original we get
+                // false positives:
+                wsMsg.resourceState =
+                    changeItem.resourceObject.state;
+                ws.send(JSON.stringify(wsMsg));
+            }
+        }
+
+        cfDrain();
```

Comment:
For the same reason it's here w/o anybody reporting issues: https://github.com/joyent/sdc-net-agent/blob/master/lib/net-agent.js#L435-L443.

My understanding is that it's possible a race condition causing getting some flood of messages right at connecting.

Reply:
Also, this is the reason for the name "cfDrain" :-)

Reply:
Ah, in the net-agent case I think it's because the changefeed listener has already been initialized in another JS event loop/cycle - and net-agent is just ensuring that it hasn't missed any 'readable' events (so it runs just to check if there has been readable events previously).

This check is not needed in this case - as everything is created/initialized in the same JS event cycle (linear).